src/lib/reasoners/relation.ml

```diff
@@ -103,26 +119,21 @@ let assume_th_elt env th_elt dep =
   let env7 = Rel7.assume_th_elt env.r7 th_elt dep in
   {r1=env1; r2=env2; r3=env3; r4=env4; r5=env5; r6=env6; r7=env7}
 
+let try_query (type a) (module R : Sig_rel.RELATION with type t = a) env uf a
+    k =
+  match Timers.with_timer R.timer Timers.F_query
+    @@ fun () -> R.query env uf a with
+  | Some r -> Some r
+  | None -> k ()
+
 let query env uf a =
-  Options.exec_thread_yield ();
-  match Rel1.query env.r1 uf a with
-  | Some _ as ans -> ans
-  | None ->
-    match Rel2.query env.r2 uf a with
-    | Some _ as ans -> ans
-    | None ->
-      match Rel3.query env.r3 uf a with
-      | Some _ as ans -> ans
-      | None ->
-        match Rel4.query env.r4 uf a with
-        | Some _ as ans -> ans
-        | None ->
-          match Rel5.query env.r5 uf a with
-          | Some _ as ans -> ans
-          | None ->
-            match Rel6.query env.r6 uf a with
-            | Some _ as ans -> ans
-            | None -> Rel7.query env.r7 uf a
+  try_query (module Rel1) env.r1 uf a @@ fun () ->
```

Comment:
I am not sure it actually matters but let's keep the `Options.exec_thread_yield ()`.

Reply:
oups! `exec_thread_yield` was here for the GUI but I agree we have to keep it as it tells use it's a hot path.